src/Lean/Parser/Term.lean

```diff
@@ -448,6 +448,10 @@ def withAnonymousAntiquot := leading_parser
 @[builtin_term_parser] def «trailing_parser» := leading_parser:leadPrec
   "trailing_parser" >> optExprPrecedence >> optExprPrecedence >> ppSpace >> termParser
 
+/-- Marks a function argument of an `@[extern]`ed function as borrowed.
+
+See https://lean-lang.org/lean4/doc/dev/ffi.html#borrowing;
+currently this annotation is ignored on `@[export]`ed functions and return values. -/
```

Comment:
Is this revision OK?

```suggestion
/-- 
Indicates that an argument to a function marked `@[extern]` is borrowed.

Being borrowed only affects the way C code interacts with the function. From the perspective of Lean, this annotation has no effect. It similarly has no effect on functions not marked `@[extern]`.

When a function argument is borrowed, the function does not consume the value. This means that the function will not decrement the value's reference count or deallocate it, and the caller is responsible for doing so.

Please see https://lean-lang.org/lean4/doc/dev/ffi.html#borrowing for a complete description.
-/
```

Reply:
@david-christiansen I would suggest a slight change:

```suggestion
/-- 
Indicates that an argument to a function is borrowed.

Being borrowed only affects how the compiler interacts with the function. From the perspective of Lean, this annotation has no effect. For the current compiler, it similarly has no effect on functions not marked `@[extern]`.

When a function argument is borrowed, the function does not consume the value. This means that the function will not decrement the value's reference count or deallocate it, and the caller is responsible for doing so.

Please see https://lean-lang.org/lean4/doc/dev/ffi.html#borrowing for a complete description.
-/
```

There are two things to note here. First, the borrow annotation does not just effect C code, it also effects IR. It is even possible to provide IR directly to an `@[extern]` definition without external C code. For example, I do this in [Lyre](https://github.com/tydeu/lyre). Second, user code can make use of the borrow annotation for more than just `@[extern]`, so I would suggest describing its general semantic meaning first, and then the way it is used by core code second.

Reply:
> Second, user code can make use of the borrow annotation for more than just @[extern].

No, AFAIK the annotation does not work on user code that is not @[extern] (for incidental reasons, even though the semantics still make sense on user code). Do you have evidence to the contrary?

Reply:
@digama0 Sorry, I was not clear. By "user code", I meant metaprogramming user code. The borrow annotation is recorded in the IR, and user code interacting with the IR (or with term syntax) can make use of that information. As you say, the semantics can still make sense in user code and some user-defined metaprogramming can take advantage of that. Thus, it seems worth to describe the semantic meaning first then describe the limitations of the core implementation.

Reply:
To be fair, that's just saying that `@&` is an annotation and therefore leaves a trace in the term. The exact same thing is true about all other annotations, and user code can do whatever it wants with annotations. Unless there are specific well known standard metaprograms that also do something with this annotation, I think it is best to stick to the description of what its actual effects are, since the part about it being an annotation which user code (potentially, but not in practice) could do something with is implicit and obviously the case.

Reply:
I've taken David's suggestion here. Mac, if you want your version, can you discuss this with David and push as you see fit?